/jds/bin/diff -uprN nautilus-2.26.3.old/libnautilus-private/apps_nautilus_preferences.schemas.in nautilus-2.26.3/libnautilus-private/apps_nautilus_preferences.schemas.in
--- nautilus-2.26.3.old/libnautilus-private/apps_nautilus_preferences.schemas.in	2009-05-18 20:44:26.000000000 +0100
+++ nautilus-2.26.3/libnautilus-private/apps_nautilus_preferences.schemas.in	2009-07-13 11:38:13.523426000 +0100
@@ -1122,6 +1122,22 @@ most cases, this should be left alone. -
          </long>
       </locale>
     </schema>
+
+    <schema>
+      <key>/schemas/apps/nautilus/desktop/primary_screen</key>
+      <applyto>/apps/nautilus/desktop/primary_screen</applyto>
+      <owner>nautilus</owner>
+      <type>int</type>
+      <default>0</default>
+      <locale name="C">
+         <short>Primary magnification screen</short>
+         <long>
+         This screen can be set if you want to have a screen other
+         than screen 0 as the default screen, while working with
+         magnifier.
+         </long>
+      </locale>
+    </schema>
         
   </schemalist>  
 </gconfschemafile>
/jds/bin/diff -uprN nautilus-2.26.3.old/libnautilus-private/nautilus-global-preferences.c nautilus-2.26.3/libnautilus-private/nautilus-global-preferences.c
--- nautilus-2.26.3.old/libnautilus-private/nautilus-global-preferences.c	2009-05-18 20:44:26.000000000 +0100
+++ nautilus-2.26.3/libnautilus-private/nautilus-global-preferences.c	2009-07-13 11:38:13.524511000 +0100
@@ -518,6 +518,10 @@ static const PreferenceDefault preferenc
 	  NULL,
 	  default_network_link_name, g_free,
 	},
+	{ NAUTILUS_PREFERENCES_DESKTOP_PRIMARY_SCREEN,
+	  PREFERENCE_INTEGER,
+	  GINT_TO_POINTER (0)
+	},
 
 	{ NAUTILUS_PREFERENCES_MEDIA_AUTOMOUNT,
 	  PREFERENCE_BOOLEAN,
/jds/bin/diff -uprN nautilus-2.26.3.old/src/file-manager/fm-icon-view.c nautilus-2.26.3/src/file-manager/fm-icon-view.c
--- nautilus-2.26.3.old/src/file-manager/fm-icon-view.c	2009-05-18 20:44:27.000000000 +0100
+++ nautilus-2.26.3/src/file-manager/fm-icon-view.c	2009-07-13 11:38:13.527520000 +0100
@@ -441,7 +441,8 @@ static gboolean
 should_show_file_on_screen (FMDirectoryView *view, NautilusFile *file)
 {
 	char *screen_string;
-	int screen_num;
+        char default_screen_string[8];
+        int screen_num, default_screen_num;
 	FMIconView *icon_view;
 	GdkScreen *screen;
 
@@ -450,10 +451,21 @@ should_show_file_on_screen (FMDirectoryV
 	if (!fm_directory_view_should_show_file (view, file)) {
 		return FALSE;
 	}
+
+        /* Get the primary magnification screen in multi-head setup */
+        if (icon_view->details->num_screens > 1) {
+                default_screen_num = eel_preferences_get_integer (NAUTILUS_PREFERENCES_DESKTOP_PRIMARY_SCREEN);
+ 
+        /* single head setup. So, default screen is 0 */
+        } else {
+                default_screen_num = 0;
+        }
+ 
+        sprintf (default_screen_string, "%d", default_screen_num);
 	
 	/* Get the screen for this icon from the metadata. */
 	screen_string = nautilus_file_get_metadata
-		(file, NAUTILUS_METADATA_KEY_SCREEN, "0");
+		(file, NAUTILUS_METADATA_KEY_SCREEN, default_screen_string);
 	screen_num = atoi (screen_string);
 	g_free (screen_string);
 	screen = gtk_widget_get_screen (GTK_WIDGET (view));
diff -ruN nautilus-2.27.91.orig/libnautilus-private/nautilus-global-preferences.h nautilus-2.27.91/libnautilus-private/nautilus-global-preferences.h
--- nautilus-2.27.91.orig/libnautilus-private/nautilus-global-preferences.h	2009-08-25 14:14:15.641559438 +0100
+++ nautilus-2.27.91/libnautilus-private/nautilus-global-preferences.h	2009-08-25 14:15:25.500286753 +0100
@@ -227,6 +227,7 @@
 #define NAUTILUS_PREFERENCES_DESKTOP_VOLUMES_VISIBLE	   "desktop/volumes_visible"
 #define NAUTILUS_PREFERENCES_DESKTOP_NETWORK_VISIBLE       "desktop/network_icon_visible"
 #define NAUTILUS_PREFERENCES_DESKTOP_NETWORK_NAME          "desktop/network_icon_name"
+#define NAUTILUS_PREFERENCES_DESKTOP_PRIMARY_SCREEN        "desktop/primary_screen"
 
 /* Lockdown */
 #define NAUTILUS_PREFERENCES_LOCKDOWN_COMMAND_LINE         "/desktop/gnome/lockdown/disable_command_line"
